design/one-pager-stacks-security-isolation.md

```diff
@@ -241,13 +246,212 @@ permissionScope: Cluster
+
+The verbs supported in these roles(admin,edit,view) will model based on existing logic around
+[built-in kubernetes roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings).
+
+|ClusterRole Name | Permissions |
+|:-------|:----------|
+| crossplane:admin | Admin: environment, namespaces, ClusterRole, Role, ClusterRoleBinding, RoleBinding |
+| crossplane:env-admin | Manage ClusterStacks, ... |
+| crossplane:env-edit  | Manage Environment Resources, ... |
+| crossplane:env-viewer | View Environment Resources, Not permitted: secrets |
+
+Roles for a foo namespace:
+ * crossplane:ns:foo:admin
+ * crossplane:ns:foo:edit
+ * crossplane:ns:foo:view
+
+To grant jane user admin access to the foo namespace, the following yaml would suffice
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+# This role binding allows "jane" to admin the Crossplane foo namespaces.
+kind: RoleBinding
+metadata:
+  name: admin-namespace
+  namespace: foo
+subjects:
+- kind: User
+  name: jane # Name is case sensitive
+  apiGroup: rbac.authorization.k8s.io
+roleRef:
+  kind: ClusterRole
+  name: crossplane:foo:admin
+  apiGroup: rbac.authorization.k8s.io
+```
+
+#### Managing CRD Permissions of Dynamically (Aggregation)
+
+As we install new stacks and new CRDs we will need to grant our built in roles access to delete, edit, create, and view
+ the given newly installed resources, and we can use the built in
+ [aggregated cluster roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles)
+ for this exact purpose.
+
+#### Aggregation Examples
+
+The built in crossplane-admin `ClusterRole` object
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane-admin
+  labels:
+    rbac.crossplane.io/role-environment: "true" # For discoverability for tools
+aggregationRule:
+  clusterRoleSelectors:
+  - matchLabels:
+      rbac.crossplane.io/aggregate-to-crossplane-admin: "true"
+rules: [] # Rules are automatically filled in by the controller manager.
+```
+
+An example of a namespace foo `ClusterRole` object
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:ns:foo:views
+  labels:
+    rbac.crossplane.io/role-namespace: "true" # For discoverability for tools
+aggregationRule:
+  clusterRoleSelectors:
+  - matchLabels:
+      rbac.crossplane.io/aggregate-to-namespace-view: "true"
+      foo: "true" // Match installed stacks for my namespace
+rules: [] # Rules are automatically filled in by the controller manager.
+```
+
+On install of crossplane-aws stack, the stack-manager would create a ClusterRoles for each built in role. In the case of
+ crossplane-admin it would contain full access to all CRDs in the stack and look like the following.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: cossplane:crossplane:stack-aws:1.0 # Unique by namespace:stack-name:version
```

Comment:
services, pods, namespaces, and many kubernetes resources are dns resolvable by their given name as part of kubernetes machinery - https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/identifiers.md

However internal objects in kubernetes seem to be more relaxed about naming. I can't find documentation on what the rules around this are. 

As a note, because ":" are allowed in a ClusterRole, but not in namespaces, by using a ":" you can be assured by convention that a malicious actor couldn't collide with your naming scheme.

As a note, looking at kubernetes roles the user facing ones don't include ":" like view, edit, admin, cluster-admin. 

We might consider following this convention for our aggregated, but user facing roles:
```
crossplane-admin
crossplane-env-admin
crossplane-env-edit
crossplane-env-view
crossplane-ns-foo-admin
...
```

``` 
> kubectl -n kube-system create clusterrole crossplane:foo:baz --verb=get --resource=pod
clusterrole "crossplane:foo:baz" created
> kubectl -n kube-system get clusterrole crossplane:foo:baz -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  creationTimestamp: 2019-10-31T18:15:42Z
  name: crossplane:foo:baz
  resourceVersion: "361976"
  selfLink: /apis/rbac.authorization.k8s.io/v1/clusterroles/crossplane%3Afoo%3Abaz
  uid: a16a0ef9-6dab-4905-984b-73eb5149940b
rules:
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
```